plugins/suspendmanager.cpp

```diff
@@ -0,0 +1,822 @@
+        for (auto npos : supportbld_would_support | transform(around(pos)))
+            if (tileHasSupportBuilding(npos)) return false;
+
+        return true;
+    }
+
+    void suspendDeadend (color_ostream &out, df::job* job) {
+        auto building = Job::getHolder(job);
+        if (!building) return;
+        coord pos(building->centerx,building->centery,building->z);
+
+        for (size_t count = 0; count < 1000; ++count){
+
+            df::building* exit = nullptr;
+            for (auto npos : neighbors | transform(around(pos))) {
+                if (!isSuitableAccess(npos))
+                    continue; // non walkable neighbour, not an exit
+
+                auto impassiblePlan = plannedImpassibleAt(npos);
+                if (!impassiblePlan)
+                    // walkable neighbour with no building scheduled, not in a dead end
+                    return;
+
+                if (leadsToDeadend.contains(impassiblePlan->id))
+                    continue; // already visited, not an exit
+
+                if (exit)
+                    return; // more than one exit, not in a dead end
+
+                exit = impassiblePlan;
+            }
+
+            if (!exit) return;
+
+            // exit is the single exit point of this corridor, suspend its construction job...
+            for (auto exit_job : exit->jobs) {
+                if (exit_job->job_type == df::job_type::ConstructBuilding) {
+                    suspensions[exit_job->id] = Reason::DEADEND;
+                }
+            }
+            // ...mark the current tile of the corridor as leading to a dead-end...
+            leadsToDeadend.insert(building->id);
+
+            // ...and continue the exploration from its position
+            building = exit;
+            pos = coord(building->centerx,building->centery,building->z);
+        }
+    }
+
+    void preserveDesigations (df::job* job){
+        CHECK_NULL_POINTER(job)
+        if (job->job_type == df::job_type::CarveTrack ||
+            job->job_type == df::job_type::SmoothFloor ||
+            job->job_type == df::job_type::DetailFloor)
+        {
+            auto building = Buildings::findAtTile(job->pos);
+            if (building) {
+                for (auto building_job : building->jobs) {
+                    if (building_job->job_type == df::job_type::ConstructBuilding) {
+                        suspensions[building_job->id] = Reason::ERASE_DESIGNATION;
+                    }
+                }
+            }
+        }
+    }
+
+    // try to fetch suspension reason for a job without modifying the map
+    // (std::map::operator[] inserts the key if it isn't found)
+    bool tryGetReason (df::job* job, Reason &reason) {
+        auto reason_it = suspensions.find(job->id);
+        if (reason_it != suspensions.end()) {
+                reason = reason_it->second;
+                return true;
+        }
+        return false;
+    }
+
+    std::map<int,Reason> suspensions;
```

Comment:
`std::unordered_map`/`std::unordered_set`? Given how often these are accessed, the performance bump could be significant

Reply:
I didn't think too hard about these, because they are expected to be small (<100 elements) unless someone is designating a mega-structure in one go. But if the consensus is that even for small maps `unordered_map` outperforms `map`, I'm happy to change that.

Reply:
Some benchmarks shows it being faster with as low as 5-10 items, I would not expect it to hurt in any case, and poooossibly make a perceptible difference in a mega-project going in the thousands of suspension. TBH the most likely scenario is that there is no perceptible difference either way, so not a big deal either way.

Reply:
Already changed it. :+1: